File: leetcode/p00219/p00219.py
from typing import List

class Solution:
    # Time Complexity: O(n)
    # Space Complexity: O(n)
    def containsNearbyDuplicate(self, nums: List[int], k: int) -> bool:
        if len(nums) <= 1:
            return False
        
        seen = {}

        for idx, num in enumerate(nums):
            if num in seen and idx - seen[num] <= k:
                return True
            seen[num] = idx
        
        return False

if __name__ == '__main__':
    s = Solution()

    assert s.containsNearbyDuplicate([1,2,3,1], 3) == True
    assert s.containsNearbyDuplicate([1,0,1,1], 1) == True
    assert s.containsNearbyDuplicate([1,2,3,1,2,3], 2) == False

File: leetcode/p00219/p00219.go
package p00219

// Time Complexity: O(n)
// Space Complexity: O(n)
func containsNearbyDuplicate(nums []int, k int) bool {
	if len(nums) <= 1 {
		return false
	}

	seen := make(map[int]int, len(nums))

	for idx, num := range nums {
		if idx2, exists := seen[num]; exists {
			if idx-idx2 <= k {
				return true
			}
		}
		seen[num] = idx
	}

	return false
}
